itou/www/apply/views/submit_views.py

```diff
@@ -348,23 +348,25 @@ def step_application(request, siae_pk, template_name="apply/submit_step_applicat
         if job_seeker.job_applications.filter(to_siae=siae).created_in_past(seconds=10).exists():
             return HttpResponseRedirect(next_url)
 
-        sender_prescriber_organization_pk = session_data.get("sender_prescriber_organization_pk")
-        sender_siae_pk = session_data.get("sender_siae_pk")
         job_application = form.save(commit=False)
         job_application.job_seeker = job_seeker
-        job_application.sender = get_user_model().objects.get(pk=session_data["sender_pk"])
+
+        # 404 is handled in atomic transaction
+        job_application.sender = get_object_or_404(User, pk=session_data["sender_pk"])
         job_application.sender_kind = session_data["sender_kind"]
-        if sender_prescriber_organization_pk:
-            job_application.sender_prescriber_organization = PrescriberOrganization.objects.get(
-                pk=sender_prescriber_organization_pk
+        if sender_prescriber_organization_pk := session_data.get("sender_prescriber_organization_pk"):
+            job_application.sender_prescriber_organization = get_object_or_404(
+                PrescriberOrganization, pk=sender_prescriber_organization_pk
             )
-        if sender_siae_pk:
-            job_application.sender_siae = Siae.objects.get(pk=sender_siae_pk)
+        if sender_siae_pk := session_data.get("sender_siae_pk"):
+            job_application.sender_siae = get_object_or_404(Siae, pk=sender_siae_pk)
         job_application.to_siae = siae
-        job_application.save()
 
-        for job in form.cleaned_data["selected_jobs"]:
-            job_application.selected_jobs.add(job)
+        with transaction.atomic():
```

Comment:
C'est sensé être assuré par `ATOMIC_REQUESTS`, pourquoi l'ajouter explicitement ? https://github.com/betagouv/itou/blob/fc0cb2086b097335bf18bf444e7c6b2c08127453/config/settings/base.py#L177

Reply:
Parce que j'ai bien cherché si ce settings (inactif par défaut) n'était pas actif et que j'ai réussi à passer à côté !
Donc effectivement mon changement est bien inutile.

Je trouve dommage que nous ayons activé `ATOMIC_REQUESTS`.

Reply:
> Je trouve dommage que nous ayons activé `ATOMIC_REQUESTS`.

Ah tiens, je veux bien que tu m'expliques pourquoi ?

Reply:
Comme dit dans la doc : 

> Avertissement

> Bien que la simplicité de ce modèle de transaction est attrayant, il devient inefficace lorsque le trafic augmente. L’ouverture d’une transaction pour chaque vue présente un certain coût. L’impact sur la performance dépend de la manière dont vos applications font usage des requêtes et de la manière dont la base de données gère les verrous.

On a par exemple une transaction ouverte dans notre appel sync à PDFShift donc mieux vaut ne pas faire le moindre update/save ici...

Dédicace à myk pour son boulot sur Django 1.6 en modifiant le fonctionnement interne de psycopg et éviter d'avoir une ouverture de transaction sur un single `.save()` et à la possibilité de traiter des requêtes HTTP avec SQL sans le moindre `BEGIN`.

Reply:
Ok compris et discuté ce matin. Merci.

Effectivement c'est probablement *wasteful* si un `BEGIN` commence avant chaque vue, même les vues qui ne feraient aucune opération en DB.

Si jamais ça se confirme on fera un point tech pour s'assurer que tout le monde comprend les transactions et on les gèrera à la main.

Reply:
La doc est ici pour ceux qui veulent lire https://docs.djangoproject.com/en/3.1/topics/db/transactions/

Il y a aussi le `non_atomic_requests` si on estime que conserver `ATOMIC_REQUESTS` en vaut la peine.

Reply:
De ce que j'ai testé le begin commence **que sur une opération de CRUD** (je vais relire des docs sur le comportement de `ATOMIC_REQUESTS` parce que ce n'est pas encore clair pour moi à quel moment il envoie le `COMMIT` ni le celui d'un select for update.
Par contre, j'ai découvert un autre truc inquiétant, je creuse avant d'en parler...

Reply:
>Par contre, j'ai découvert un autre truc inquiétant, je creuse avant d'en parler...

Quel suspense @stephane ! Faut pas me laisser dans l'expectative comme ça ! :smile: